src/plugins/intel_cpu/tests/functional/subgraph_tests/src/stateful_model.cpp

```diff
@@ -0,0 +1,547 @@
+            ASSERT_TRUE(outputTensor);
+            inferRequest.infer();
+            std::vector<float> expected_state0;
+            std::vector<float> expected_results;
+            std::tie(expected_state0, expected_results) = calc_refs();
+
+            auto states = inferRequest.query_state();
+            ASSERT_FALSE(states.empty());
+            ov::Tensor state0;
+            ov::Tensor state1;
+            for (auto&& state : states) {
+                if ("variable0" == state.get_name()) {
+                    state0 = state.get_state();
+                }
+                if ("variable1" == state.get_name()) {
+                    state1 = state.get_state();
+                }
+            }
+            ASSERT_TRUE(state0);
+            ASSERT_TRUE(state1);
+            auto actual_result = outputTensor.data<ov::element_type_traits<testPrc>::value_type>();
+            float_compare(&expected_state0[i], state0.data<ov::element_type_traits<testPrc>::value_type>(), 1);
+            float_compare(&expected_results[i], state1.data<ov::element_type_traits<testPrc>::value_type>(), 1);
+            float_compare(&expected_results[i], actual_result, 1);
+        }
+    }
+};
+
+TEST_F(StaticShapeTwoStatesModel, smoke_Run_Static_Two_States) {
+    prepare();
+    run_test();
+}
+
+// ┌─────────┐    ┌───────────┐
+// │ Param1  │--->│ ReadValue │..
+// └───┬──┬──┘    └─────┬─────┘ .
+//     │  │             │       .
+//     │  │             │       .
+//     │  └──┬─────┬────┘       .
+//     │     │ Add │            .
+//     │     └──┬──┘            .
+//     │        │               .
+//     │    ┌───┴────┐          .
+//     └────┤ Concat │          .
+//          └────────┘          .
+//             / \              .
+//            /   \             .
+//   ┌────────┐   ┌────────┐    .
+//   │ Result │   │ Assign │.....
+//   └────────┘   └────────┘
+
+class DynamicShapeStatefulModel : public StatefulModelTest {
+public:
+    void SetUp() override {
+        targetDevice = ov::test::utils::DEVICE_CPU;
+        ov::element::Type netPrc = testPrc;
+
+        const ov::Shape inpShape = {1, 1};
+        const InputShape input_shape = {{-1, 1}, {{1, 1}, {2, 1}, {4, 1}, {8, 1}, {16, 1}}};
+        init_input_shapes({input_shape});
+
+        auto arg = std::make_shared<ov::op::v0::Parameter>(netPrc, inputDynamicShapes.front());
+
+        // The ReadValue/Assign operations must be used in pairs in the model.
+        // For each such a pair, its own variable object must be created.
+        const std::string variable_name("variable0");
+        auto variable = std::make_shared<ov::op::util::Variable>(
+            ov::op::util::VariableInfo{ov::PartialShape::dynamic(), ov::element::dynamic, variable_name});
+
+        // Creating ov::Model
+        auto read = std::make_shared<ov::op::v6::ReadValue>(arg, variable);
+        std::vector<std::shared_ptr<ov::Node>> args = {arg, read};
```

Comment:
Do we expect a behavior difference of the two lines below?
`auto read = std::make_shared<ov::op::v6::ReadValue>(arg, variable);`
and
`auto read = std::make_shared<ov::op::v6::ReadValue>(variable);`   (which is newly added by pull/20404)

Should the first object be initialized by `arg`? If so, the reference state output after the first iteration should be concat(2.44f, add(2.44f, 2.44f)), instead of concat(2.44f, add(2.44f, 0.f)).

@itikhono Would you please comment on the behavior of state with a initial state from a subgraph?

Reply:
To my understanding, in this particular tests, the behavior is the same. On L340 `states.front().set_state(init_tensor);` with 0 value is explicitly called, thus we always start with `concat(2.44f, add(2.44f, 0.f))`.